From 6e06b6d4ab35f5c972cf12a15b97ddcd53052d5f Mon Sep 17 00:00:00 2001
From: igo95862 <igo95862@yandex.ru>
Date: Sat, 5 Aug 2023 21:24:39 +0600
Subject: [PATCH 3/5] Ensure that bubblejail sandbox helper does not treat
 passed arguments as options

Apparently `REMAINDER` of the argparse module is undocumented as of
Python 3.9 and its use is discouraged.

https://github.com/python/cpython/issues/87509

Instead switch arguments to run to the `nargs="*"` and make
bubblejail pass the option terminating `--` to the helper. This way
helper will not treat passed `--foo` arguments as options.
---
 src/bubblejail/bubblejail_helper.py |  3 +--
 src/bubblejail/bubblejail_runner.py |  2 ++
 test/test_helper.py                 | 16 ++++++----------
 3 files changed, 9 insertions(+), 12 deletions(-)

diff --git a/src/bubblejail/bubblejail_helper.py b/src/bubblejail/bubblejail_helper.py
index e37261d..ebdc0bb 100644
--- a/src/bubblejail/bubblejail_helper.py
+++ b/src/bubblejail/bubblejail_helper.py
@@ -15,7 +15,6 @@
 # along with bubblejail.  If not, see <https://www.gnu.org/licenses/>.
 from __future__ import annotations
 
-from argparse import REMAINDER as ARG_REMAINDER
 from argparse import ArgumentParser
 from asyncio import (
     CancelledError,
@@ -446,7 +445,7 @@ def get_helper_argument_parser() -> ArgumentParser:
 
     parser.add_argument(
         'args_to_run',
-        nargs=ARG_REMAINDER,
+        nargs="*",
     )
 
     return parser
diff --git a/src/bubblejail/bubblejail_runner.py b/src/bubblejail/bubblejail_runner.py
index 5f77c32..2dd7343 100644
--- a/src/bubblejail/bubblejail_runner.py
+++ b/src/bubblejail/bubblejail_runner.py
@@ -288,6 +288,8 @@ class BubblejailRunner:
         if self.is_shell_debug:
             yield "--shell"
 
+        yield "--"
+
     def read_info_fd(self) -> None:
         with open(self.info_fd_pipe_read, closefd=False) as f:
             info_dict = json_load(f)
diff --git a/test/test_helper.py b/test/test_helper.py
index 299ec93..6e6e35c 100644
--- a/test/test_helper.py
+++ b/test/test_helper.py
@@ -92,13 +92,13 @@ class HelperParserTests(TestCase):
 
     def test_parser(self) -> None:
         """Test how helper argument parser works"""
-        required_args = ['--helper-socket', '0']
+        required_args = ['--helper-socket', '0', '--']
 
         with self.subTest('No shell'):
             no_shell_example_args = [
                 '/bin/true',
                 '--long-opt', '-e', '-test',
-                '/bin/false', '--shell'
+                '/bin/false', '--shell',
             ]
             parsed_args = self.parser.parse_args(
                 required_args + no_shell_example_args
@@ -109,24 +109,20 @@ class HelperParserTests(TestCase):
 
         with self.subTest('Shell plus args'):
             with_shell_example_args = [
-                '--shell', '/bin/ls', '-l'
+                 '/bin/ls', '-l'
             ]
 
             parsed_args = self.parser.parse_args(
-                required_args + with_shell_example_args
+                ['--shell'] + required_args + with_shell_example_args
             )
 
             self.assertTrue(parsed_args.shell)
             self.assertEqual(
-                parsed_args.args_to_run, with_shell_example_args[1:])
+                parsed_args.args_to_run, with_shell_example_args)
 
         with self.subTest('Only shell'):
-            just_shell_example = [
-                '--shell'
-            ]
-
             parsed_args = self.parser.parse_args(
-                required_args + just_shell_example
+                ['--shell'] + required_args
             )
 
             self.assertTrue(parsed_args.shell)
-- 
2.41.0

